File: kageyomi/uavp/receipt_manager.py
import json
import hashlib
import os
import httpx
from eth_account.messages import encode_defunct
from eth_account import Account

# ensure that we use the unencrypted key locally
def normalize_hex(val: str) -> str:
    return val if val.startswith("0x") else f"0x{val}"

def sha256_hex(val: str | bytes) -> str:
    if isinstance(val, str):
        val = val.encode("utf-8")
    return "0x" + hashlib.sha256(val).hexdigest()

def create_receipt(job_id: str, tool_name: str, tool_call_id: str, tool_args_json: str, params_hash: str, response_hash: str, canonical_data: str, data_type: str, timestamp_ms: int) -> dict:
    obj = {
        "jobId": job_id,
        "toolName": tool_name,
        "toolCallId": tool_call_id,
        "toolArgumentsJson": tool_args_json,
        "paramsHash": params_hash,
        "responseHash": response_hash,
        "dataType": data_type,
        "canonicalData": canonical_data,
        "timestampMs": timestamp_ms
    }
    receipt_hash = sha256_hex(json.dumps(obj, sort_keys=True, separators=(",", ":")))
    return {**obj, "receiptHash": receipt_hash}

def sign_receipt(receipt: dict, private_key: str) -> dict:
    if not private_key:
        return receipt
    msg = encode_defunct(hexstr=receipt["receiptHash"])
    signed = Account.sign_message(msg, private_key=normalize_hex(private_key))
    return {**receipt, "signature": signed.signature.hex()}


def _local_receipt_cache_dir() -> str:
    return os.getenv("KAGEYOMI_RECEIPT_CACHE_DIR", ".receipt-cache")


def _persist_receipts_locally(cid: str, receipts: list) -> None:
    cache_dir = _local_receipt_cache_dir()
    os.makedirs(cache_dir, exist_ok=True)
    with open(os.path.join(cache_dir, f"{cid}.json"), "w", encoding="utf-8") as handle:
        json.dump(receipts, handle, indent=2)

async def post_receipts_to_ipfs(receipts: list) -> str:
    if os.getenv("KAGEYOMI_MOCK_IPFS", "false").lower() == "true":
        cid = sha256_hex(json.dumps(receipts))[2:]
        mock_dir = os.getenv("KAGEYOMI_MOCK_IPFS_DIR", ".mock-ipfs")
        os.makedirs(mock_dir, exist_ok=True)
        with open(os.path.join(mock_dir, f"{cid}.json"), "w", encoding="utf-8") as f:
            json.dump(receipts, f, indent=2)
        _persist_receipts_locally(cid, receipts)
        return cid

    pinata_jwt = os.getenv("PINATA_JWT")
    if not pinata_jwt:
        raise ValueError("PINATA_JWT is not configured")

    payload = json.dumps(receipts).encode("utf-8")
    files = {"file": ("kageyomi-receipts.json", payload)}
    data = {"network": "public", "name": "kageyomi-receipts.json"}
    
    async with httpx.AsyncClient() as client:
        res = await client.post(
            "https://uploads.pinata.cloud/v3/files",
            headers={"Authorization": f"Bearer {pinata_jwt}"},
            data=data,
            files=files,
            timeout=30.0
        )
        res.raise_for_status()
        resp_data = res.json()
        cid = resp_data.get("data", {}).get("cid")
        if not cid:
            raise ValueError(f"IPFS receipt upload returned no CID: {resp_data}")
        _persist_receipts_locally(cid, receipts)
        return cid


async def load_receipts_from_ipfs(cid: str) -> list[dict]:
    local_path = os.path.join(_local_receipt_cache_dir(), f"{cid}.json")
    if os.path.exists(local_path):
        with open(local_path, "r", encoding="utf-8") as handle:
            return json.load(handle)

    mock_dir = os.getenv("KAGEYOMI_MOCK_IPFS_DIR", ".mock-ipfs")
    mock_path = os.path.join(mock_dir, f"{cid}.json")
    if os.path.exists(mock_path):
        with open(mock_path, "r", encoding="utf-8") as handle:
            return json.load(handle)

    gateway_base = os.getenv("IPFS_GATEWAY", "https://gateway.pinata.cloud/ipfs")
    gateway_url = f"{gateway_base.rstrip('/')}/{cid}"
    async with httpx.AsyncClient(timeout=30.0) as client:
        res = await client.get(gateway_url)
        res.raise_for_status()
        receipts = res.json()
    if not isinstance(receipts, list):
        raise ValueError(f"Receipts CID {cid} did not resolve to a list payload")
    _persist_receipts_locally(cid, receipts)
    return receipts

def compute_merkle_root(leaves: list[str]) -> str:
    if not leaves:
        return sha256_hex("")
    
    level = [normalize_hex(l) for l in leaves]
    while len(level) > 1:
        next_level = []
        for i in range(0, len(level), 2):
            left = level[i]
            right = level[i+1] if i+1 < len(level) else left
            next_level.append(sha256_hex(bytes.fromhex(left[2:]) + bytes.fromhex(right[2:])))
        level = next_level
    return level[0]

def compute_trace_hash(prompt: str, output_hash: str, receipt_root: str, receipts_cid: str, model: str, tool_count: int) -> str:
    obj = {
        "promptHash": sha256_hex(prompt),
        "outputHash": normalize_hex(output_hash),
        "receiptRoot": normalize_hex(receipt_root),
        "receiptsCID": receipts_cid,
        "model": model,
        "toolCount": tool_count
    }
    return sha256_hex(json.dumps(obj, sort_keys=True, separators=(",", ":")))

File: apps/agent/src/uavp/receipt-manager.ts
import crypto from 'node:crypto'
import { mkdir, writeFile } from 'node:fs/promises'
import path from 'node:path'
import { privateKeyToAccount } from 'viem/accounts'

import type { CanonicalDataType } from '../canonicalize/uavp-canonicalize.js'

export type ToolReceiptInput = {
  jobId: string
  toolName: string
  toolCallId: string
  toolArgumentsJson: string
  paramsHash: string
  responseHash: string
  canonicalData: string
  dataType: CanonicalDataType
  timestampMs: number
}

export type ToolReceipt = ToolReceiptInput & {
  receiptHash: string
  signature?: string
}

export function createReceipt(input: ToolReceiptInput): ToolReceipt {
  const receiptHash = sha256Hex(
    JSON.stringify(
      {
        jobId: input.jobId,
        toolName: input.toolName,
        toolCallId: input.toolCallId,
        toolArgumentsJson: input.toolArgumentsJson,
        paramsHash: input.paramsHash,
        responseHash: input.responseHash,
        dataType: input.dataType,
        canonicalData: input.canonicalData,
        timestampMs: input.timestampMs,
      },
      Object.keys(input).sort(),
    ),
  )

  return {
    ...input,
    receiptHash,
  }
}

export async function signReceipt(receipt: ToolReceipt, privateKey: string): Promise<ToolReceipt> {
  const account = privateKeyToAccount(normalizeHex(privateKey))
  const signature = await account.signMessage({ message: { raw: hexToBytes(receipt.receiptHash) } })
  return {
    ...receipt,
    signature,
  }
}

export async function postReceiptsToIPFS(receipts: ToolReceipt[]): Promise<string> {
  if (process.env.KAGEYOMI_MOCK_IPFS === 'true') {
    const cid = sha256Hex(JSON.stringify(receipts)).slice(2)
    const directory = process.env.KAGEYOMI_MOCK_IPFS_DIR || path.resolve(process.cwd(), '.mock-ipfs')
    await mkdir(directory, { recursive: true })
    await writeFile(path.join(directory, `${cid}.json`), JSON.stringify(receipts, null, 2), 'utf8')
    return cid
  }

  const pinataJwt = process.env.PINATA_JWT
  if (!pinataJwt) {
    throw new Error('PINATA_JWT is not configured')
  }

  const payload = Buffer.from(JSON.stringify(receipts), 'utf8')
  const formData = new FormData()
  formData.append('file', new Blob([payload]), 'kageyomi-receipts.json')
  formData.append('network', 'public')
  formData.append('name', 'kageyomi-receipts.json')

  const response = await fetch('https://uploads.pinata.cloud/v3/files', {
    method: 'POST',
    headers: {
      Authorization: `Bearer ${pinataJwt}`,
    },
    body: formData,
  })

  if (!response.ok) {
    throw new Error(`IPFS receipt upload failed: ${response.status} ${response.statusText}`)
  }

  const json = (await response.json()) as { data?: { cid?: string } }
  const cid = json.data?.cid
  if (!cid) {
    throw new Error(`IPFS receipt upload returned no CID: ${JSON.stringify(json)}`)
  }

  return cid
}

export function computeMerkleRoot(leaves: string[]): string {
  if (leaves.length === 0) {
    return sha256Hex('')
  }

  let level: Array<`0x${string}`> = leaves.map((leaf) => normalizeHex(leaf))
  while (level.length > 1) {
    const next: Array<`0x${string}`> = []
    for (let index = 0; index < level.length; index += 2) {
      const left = level[index]
      const right = level[index + 1] ?? left
      next.push(normalizeHex(sha256Hex(Buffer.concat([hexToBytes(left), hexToBytes(right)]))))
    }
    level = next
  }
  return level[0]
}

export function sha256Hex(value: string | Buffer): string {
  return `0x${crypto.createHash('sha256').update(value).digest('hex')}`
}

export function computeTraceHash(input: {
  prompt: string
  outputHash: string
  receiptRoot: string
  receiptsCID: string
  model: string
  toolCount: number
}): string {
  return sha256Hex(
    JSON.stringify(
      {
        promptHash: sha256Hex(input.prompt),
        outputHash: normalizeHex(input.outputHash),
        receiptRoot: normalizeHex(input.receiptRoot),
        receiptsCID: input.receiptsCID,
        model: input.model,
        toolCount: input.toolCount,
      },
      Object.keys(input).sort(),
    ),
  )
}

function normalizeHex(value: string): `0x${string}` {
  return (value.startsWith('0x') ? value : `0x${value}`) as `0x${string}`
}

function hexToBytes(value: string): Uint8Array {
  return Uint8Array.from(Buffer.from(value.replace(/^0x/, ''), 'hex'))
}
